scripts/globals/npc_util.lua

```diff
@@ -38,175 +49,254 @@ function npcUtil.pickNewPosition(npc, positionTable, allowCurrentPosition)
+        player:messageSpecial(ITEM_OBTAINED, v[1]);
     end
+    return true;
+end
 
-    player:messageSpecial(ITEM_OBTAINED, itemId);
-
-    return true
+--[[ *******************************************************************************
+    Give key item(s) to player.
+    Message is displayed showing key items obtained.
+    
+    Examples of valid keyitems parameter:
+        ZERUHN_REPORT
+        {PALBOROUGH_MINES_LOGS}
+        {BLUE_ACIDITY_TESTER, RED_ACIDITY_TESTER}
+******************************************************************************* --]]
+function npcUtil.giveKeyItem(player, keyitems)
+    -- require zone TextIDs
+    local TextIDs = "scripts/zones/" .. player:getZoneName() .. "/TextIDs";
+    package.loaded[TextIDs] = nil;
+    require(TextIDs);
+    
+    -- create table of keyitems
+    local givenKeyItems = {};
+    if (type(keyitems) == "number") then
+        givenKeyItems = {keyitems};
+    elseif (type(keyitems) == "table") then
+        givenKeyItems = keyitems;
+    else
+        print(string.format("ERROR: invalid keyitems parameter given to npcUtil.giveKeyItem in zone %s.", player:getZoneName()));
+        return false;
+    end
+    
+    -- give key items to player, with message
+    for _, v in pairs(givenKeyItems) do
+        player:addKeyItem(v);
+        player:messageSpecial(KEYITEM_OBTAINED,v);
+    end
+    return true;
 end
 
--- Complete a quest
--- This can return false if player could not recieve items
--- Fame defaults to 30
--- all params in table are optional
---
--- npcUtil.completeQuest(player, SANDORIA, ROSEL_THE_ARMORER, {
---            title= ENTRANCE_DENIED,
---            gil= 200,
---            fame=120,
---            keyItem=100,
---             item={927, 927}
---            });
-function npcUtil.completeQuest(player, area, quest, options)
-
-    options = options or {}
-
-    if (options["item"] ~= nil) then
-        if (npcUtil.giveItem(player, options["item"]) == false) then
-            return false
+--[[ *******************************************************************************
+    Complete a quest.
+    If quest rewards items, and the player cannot carry them, return false.
+    Otherwise, return true.
+    
+    Example of usage with params (all params are optional):
+        npcUtil.completeQuest(player, SANDORIA, ROSEL_THE_ARMORER, {
+            item = { {640,2}, 641 },    -- see npcUtil.giveItem for formats
+            keyItem = ZERUHN_REPORT,    -- see npcUtil.giveKeyItem for formats
+            fame = 120,                 -- fame defaults to 30 if not set
+            bayld = 500,
+            gil = 200,
+            xp = 1000,
+            title = ENTRANCE_DENIED,
+        });
+******************************************************************************* --]]
+function npcUtil.completeQuest(player, area, quest, params)
+    params = params or {};
+
+    -- load text ids
+    local TextIDs = "scripts/zones/" .. player:getZoneName() .. "/TextIDs";
```

Comment:
shouldn't the text IDs have already been loaded by the script that called this function?

Reply:
Trying to make it new contributor-proof.  How much faith do you have in your script makers? 🎲